bundles/org.eclipse.swt/Eclipse SWT Custom Widgets/common/org/eclipse/swt/custom/BusyIndicator.java

```diff
@@ -59,32 +61,154 @@ public static void showWhile(Display display, Runnable runnable) {
 			}
 		}
 
-		Integer busyId = Integer.valueOf(nextBusyId);
-		nextBusyId++;
-		Cursor cursor = display.getSystemCursor(SWT.CURSOR_WAIT);
-		Shell[] shells = display.getShells();
-		for (Shell shell : shells) {
-			Integer id = (Integer)shell.getData(BUSYID_NAME);
-			if (id == null) {
-				setCursorAndId(shell, cursor, busyId);
-			}
-		}
+		Integer busyId = setBusyCursor(display);
 
 		try {
 			runnable.run();
 		} finally {
-			shells = display.getShells();
-			for (Shell shell : shells) {
-				Integer id = (Integer)shell.getData(BUSYID_NAME);
-				if (Objects.equals(id, busyId)) {
-					setCursorAndId(shell, null, null);
+			clearBusyCursor(display, busyId);
+		}
+	}
+
+	/**
+	 * If called from a {@link Display} thread use the given {@link SwtRunnable} to
+	 * produces a {@link CompletableFuture} providing busy feedback using the busy
+	 * indicator while execution is running. If called from a non {@link Display}
+	 * the execution is performed in place and the result returned as a
+	 * {@link CompletableFuture}. It is therefore safe to call this method from any
+	 * thread and the {@link SwtCallable} is always evaluated outside the UI. The
+	 * {@link ForkJoinPool#commonPool()} is used to execute the computation in case
+	 * this is called from a {@link Display} thread
+	 *
+	 * @param action the action that should be executed and produces a result of the
+	 *               {@link CompletableFuture}
+	 *
+	 * @since 3.122
+	 */
+	public static <E extends Exception> CompletableFuture<?> execute(SwtRunnable<E> action) {
+		return execute(action, ForkJoinPool.commonPool());
+	}
+
+	/**
+	 * If called from a {@link Display} thread use the given {@link SwtRunnable} to
+	 * produces a {@link CompletableFuture} providing busy feedback using the busy
+	 * indicator while execution is running. If called from a non {@link Display}
+	 * the execution is performed in place and the result returned as a
+	 * {@link CompletableFuture}. It is therefore safe to call this method from any
+	 * thread and the {@link SwtCallable} is always evaluated outside the UI.
+	 *
+	 * @param action   the action that should be executed and produces a result of
+	 *                 the {@link CompletableFuture}
+	 * @param executor the Executor to perform the computation in case this is
+	 *                 called from a {@link Display} thread, passing a
+	 *                 {@link Display} will throw an
+	 *                 {@link IllegalArgumentException} as this will lead to a
+	 *                 blocking UI and violates the contract of this method
+	 * @since 3.122
+	 */
+	public static <E extends Exception> CompletableFuture<?> execute(SwtRunnable<E> action, Executor executor) {
+		return compute(() -> {
```

Comment:
I don't the see passed Executor being used here. I'm certainly missing something.

Reply:
Ah, ok it's just missing as an argument to the delegate compute()

Reply:
Good catch, this should be fixed now!